core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -405,33 +406,33 @@ private RebalanceDecision decideRebalance(final ShardRouting shard) {
                         assignedNode = node;
                     }
                 }
-                final int nodeRanking;
-                if (rebalanceConditionsMet || betterWeightThanCurrent == false) {
-                    if (betterWeightThanCurrent == false && currentNodeWeightRanking == 0) {
-                        // current node's ranking isn't assigned yet, and we are already examining nodes that
-                        // are of worse ranking, so assign the current node's ranking first
-                        currentNodeWeightRanking = ++weightRanking;
-                    }
-                    // the node meets the conditions for being a better balance than the current for the given shard,
-                    // or the node has a worse weight than the current, so increment the weight ranking
-                    nodeRanking = ++weightRanking;
+                Tuple<ModelNode, Decision> nodeResult = Tuple.tuple(node, canAllocate);
+                if (rebalanceConditionsMet) {
+                    betterBalanceNodes.add(nodeResult);
+                } else if (betterWeightThanCurrent) {
+                    sameBalanceNodes.add(nodeResult);
                 } else {
-                    if (currentNodeWeightRanking == 0) {
-                        currentNodeWeightRanking = ++weightRanking;
-                    }
-                    nodeRanking = currentNodeWeightRanking;
+                    worseBalanceNodes.add(nodeResult);
                 }
-                nodeDecisions.put(node.getNodeId(), new NodeAllocationResult(
-                    node.routingNode.node(),
-                    rebalanceConditionsMet ? canAllocate.type() : Type.NO,
-                    canAllocate,
-                    nodeRanking)
-                );
             }
 
-            if (currentNodeWeightRanking == 0) {
-                // the current node is the worst ranked, so its ranking never got assigned above
-                currentNodeWeightRanking = weightRanking + 1;
+            int weightRanking = 0;
+            Map<String, NodeAllocationResult> nodeDecisions = new HashMap<>(modelNodes.length - 1);
```

Comment:
do we need a map here? As we sort anyway in the `RebalanceDecision` constructor, is it good enough to just pass a list of `NodeRebalanceResult` to it?